plutus-core/untyped-plutus-core/Language/UntypedPlutusCore/Core/Instance/Flat.hs

```diff
@@ -0,0 +1,121 @@
+
+However, having this flexibility allows us to encode e.g. PLC with substantial annotations
+(like position information) in situation where the stability is *not* critical, such as
+for testing.
+-}
+
+{- Note [Encoding/decoding constructor tags using Flat]
+Flat saves space when compared to CBOR by allowing tags to use the minimum
+number of bits required for their encoding.
+
+This requires specialised encode/decode functions for each constructor
+that encodes a different number of possibilities. Here is a list of the
+tags and their used/available encoding possibilities.
+
+| Data type       | Function          | Used | Available |
+|-----------------|-------------------|------|-----------|
+| Terms           | encodeTerm        | 8    | 16        |
+
+For format stability we are manually assigning the tag values to the
+constructors (and we do not use a generic algorithm that may change this order).
+
+All encodings use the function `eBits :: NumBits -> Word8 -> Encoding`, which encodes
+at most 8 bits of data, and the first argument specifies how many bits from the 8
+available are actually used.
+
+Flat uses an extra function in its class definition called `size`. Since we want
+to reserve some space for future data constructors and we don't want to have the
+sizes desynchronised from the encoding and decoding functions we have manual
+implementations for them (if they have any constructors reserved for future use).
+-}
+
+-- | Using 4 bits to encode term tags.
+termTagWidth :: NumBits
+termTagWidth = 4
+
+encodeTerm :: Word8 -> Encoding
+encodeTerm = eBits termTagWidth
+
+decodeTerm :: Get Word8
+decodeTerm = dBEBits8 termTagWidth
+
+instance ( Closed uni
+         , uni `Everywhere` Flat
+         , Flat ann
+         , Flat name
+         ) => Flat (Term name uni ann) where
+    encode = \case
+        Var      ann n    -> encodeTerm 0 <> encode ann <> encode n
+        Delay    ann t    -> encodeTerm 1 <> encode ann <> encode t
+        LamAbs   ann n t  -> encodeTerm 2 <> encode ann <> encode n <> encode t
+        Apply    ann t t' -> encodeTerm 3 <> encode ann <> encode t <> encode t'
+        Constant ann c    -> encodeTerm 4 <> encode ann <> encode c
+        Force    ann t    -> encodeTerm 5 <> encode ann <> encode t
+        Error    ann      -> encodeTerm 6 <> encode ann
+        Builtin  ann bn   -> encodeTerm 7 <> encode ann <> encode bn
+
+    decode = go =<< decodeTerm
+        where go 0 = Var      <$> decode <*> decode
+              go 1 = Delay    <$> decode <*> decode
+              go 2 = LamAbs   <$> decode <*> decode <*> decode
+              go 3 = Apply    <$> decode <*> decode <*> decode
+              go 4 = Constant <$> decode <*> decode
+              go 5 = Force    <$> decode <*> decode
+              go 6 = Error    <$> decode
+              go 7 = Builtin  <$> decode <*> decode
+              go _ = fail "Failed to decode Term TyName Name ()"
```

Comment:
wrong type

Reply:
If you mean that the TyName is extra, then I removed it, but I don't know what else you could mean by wrong type.

Reply:
that was all!